packages/open-next/src/adapters/server-adapter.ts

```diff
@@ -210,13 +222,89 @@ function createRequestHandler() {
   }).getRequestHandler();
 }
 
-async function processRequest(req: IncomingMessage, res: ServerResponse) {
+// NOTE: As of Nextjs 13.3.13+, the middleware is handled outside the next-server.
+// OpenNext will run the middleware in a sandbox and set the appropriate req headers
+// and res.body prior to processing the next-server.
+async function handleMiddleware(
+  req: IncomingMessage,
+  res: ServerResponse,
+  rawPath: string
+) {
+  const hasMatch = middleMatch.find((r) => r.test(rawPath));
+  if (!hasMatch) return;
+
+  // NOTE: Next middleware was originally developed to support nested middlewares
+  // but that was discarded for simplicity. The MiddlewareInfo type still has the original
+  // structure, but as of now, the only useful property on it is the "/" key (ie root).
+  const middlewareInfo = middlewareManifest.middleware["/"];
+  middlewareInfo.paths = middlewareInfo.files.map((file) =>
+    path.join(NEXT_DIR, file)
+  );
+
+  const result = await run({
+    distDir: NEXT_DIR,
+    name: middlewareInfo.name || "/",
+    paths: middlewareInfo.paths || [],
+    edgeFunctionEntry: middlewareInfo,
+    request: {
+      headers: req.headers,
+      method: req.method || "GET",
+      nextConfig: {
+        basePath: config.basePath,
+        i18n: config.i18n,
+        trailingSlash: config.trailingSlash,
+      },
+      url: `http://localhost:3000${rawPath}`, // internal host
+      body: getCloneableBody(req),
+      signal: signalFromNodeResponse(res),
+    },
+    useCache: true,
+    onWarning: console.warn,
+  });
+  res.statusCode = result.response.status;
+  // If the middleware returned a Redirect, we set the `Location` header with
+  // the redirected url and end the response.
+  if (res.statusCode >= 300 && res.statusCode < 400) {
+    const location = result.response.headers
+      .get("location")
+      ?.replace("http://localhost:3000", `https://${req.headers.host}`);
+    res.setHeader("Location", location);
+    return res.end();
+  }
+  // Apply the headers from the middleware to the `req` to be processed by the server
+  for (let [key, value] of result.response.headers) {
```

Comment:
We may need to omit some headers.
TODO: double check

Reply:
There is a `x-middleware-override-headers` that the middleware set that we may need to use. I have not tested this but the middleware return the overriden headers as `x-middleware-request-HEADER_NAME`, we might need to apply them ourselves.

I did this, not sure if it's necessary: 
```typescript
const toOverride: string =
  middlewareHeaders['x-middleware-override-headers'];
const overrideHeaders = toOverride.split(',').reduce(
  (acc, cur) => ({
    ...acc,
    [cur]: middlewareHeaders[`x-middleware-request-${cur}`],
  }),
  {}
);
```

Reply:
Is there a good way to trigger that? I'm not seeing the `x-middleware-override-headers` from the `result.response.headers`.
In my middleware, I'm setting some test headers via `NextResponse.next({headers: {}})`

Reply:
My bad it seems it's only for when we modify the request headers. https://nextjs.org/docs/app/building-your-application/routing/middleware#setting-headers. We still need to do it though